### ISSUE TITLE
Setup docker CI for all repos that produce a docker image

### ISSUE THREAD

**Issue by adriancole** [hypertrace/hypertrace#51]
all repos that produce docker images should test the image(s) they built.

Since we write HEALTHCHECK always now, the most portable basic test is parsing that. This helps avoid a dud image due to simple problems like file renames.

The most simple test is running a container with a fixed name (ex `docker run --rm --name test ...`) and parsing health check status `docker inspect --format='{{json .State.Health.Status}}' test|jq .` IIRC this is already done on push when using DockerHub.

To do before that means an explicit command in github actions or circleci. We are using gradle instead of [github actions](https://github.com/docker/build-push-action) to build/push docker images. It is possible something like this could be done in gradle, or we can do something in circleci like https://github.com/hypertrace/hypertrace/pull/13, noting especially leaf repos may not need to use docker-compose.

cc @aaron-steinfeld @jcchavezs for thoughts.

**Comment by aaron-steinfeld** [comment]
Using gradle to build it is a bit independent of using github actions or circleci to invoke the command. While I'd love to move to github actions, I don't think that buys us much at this point.

So as far as running this on build, the first step is, for each image, to come up with the specific docker or docker-compose invocation. Once we've got that, I believe there are a couple gotchas as to how to run that since our build is already running in a docker container, but nothing too dramatic: https://circleci.com/docs/2.0/docker-compose/#using-docker-compose-with-docker-executor . I _think_ we just are limited on volume mounts and might have to work around that, if we need them.

**Comment by adriancole** [comment]
aside, but maybe misunderstood the docker github action as I embedded the link https://github.com/docker/build-push-action  While agree that this plugin doesn't do a healtcheck at the moment, it is easier to work with in docker than circleci which has the volume limitation and also maintenance around overriding versions etc. It definitely does buy us something to use something maintained by docker and directly integrates with github, vs using circleci and custom jobs, overrides or workarounds to the constraints.

**Comment by jcchavezs** [comment]
I think using gradle to build/push images makes things non explicit and hard to digest. Docker has a very nice API that we can use explicitly and then understand and debug. I am really curious on the reasons (and benefits) for using gradle for docker stuff.

> I think we just are limited on volume mounts and might have to work around that, if we need them.

You can't mount if the executor is docker. You can if the executor is a machine. In any case I'd suggest to use docker if we don't need to mount folders.

**Comment by adriancole** [comment]
the mounting we currently do is unnecessary, I'll sideline that

**Comment by aaron-steinfeld** [comment]
> I think using gradle to build/push images makes things non explicit and hard to digest. Docker has a very nice API that we can use explicitly and then understand and debug. I am really curious on the reasons (and benefits) for using gradle for docker stuff.

A few reasons:
- The build of the actual source code/artifacts, staging of the build context, version calculation etc. are already in gradle so keeping the image build there too means we have one entry point for all build tasks, rather than having to switch to different tools for different tasks. This means that the tasks can understand their dependencies and ordering and parallelize/sequence to optimize build times. It also means that the tasks are discoverable (`gw tasks`) and consistent between CI and local (vs a manual invocation of different tools, or using something like the native github actions commands which as far as I'm aware aren't usable locally)
- Because we can share gradle plugins, we can keep builds consistent across images without copy/pasting scripts around
- We're doing very little with docker itself in gradle, so for the trivial invocation of docker build and docker push, we're not losing much visibility (less a reason than a non-reason)

**Comment by adriancole** [comment]
fwiw the most common thing I do is `docker build . -t whatever`, later doing the corresponding gradle things to make sure that part isn't busted.

It isn't true that the gradle plugins do only a little. in some cases it is very much more complicated than a HEREFILE to do the same. They also have been very high maintenance, split between large amounts of circleci yaml, and several versions to update. It is also not idiomatic to use gradle to issue docker commands for projects that produce no java code. Even ones that do, especially once repo consolidation is done there would only be a few to several java service images.

A lot of people will be able to run and understand common tools instead of custom ones.  will have basic experience at least with normal docker and less need for these very custom things we have here vs what are norms of OSS. Using things that are harder and more likely to work for an org than will have maybe 10 docker images is a cure worse than the disease. In a different context, such as an internal thing that makes a lot of images in a custom way, it could make sense, but I really think it is not serving.

This opinion is based on working on other projects and also our projects. I have perspective and have been working with docker since the project started. It isn't conjecture.. it is a real opinion. I do as much or more work on the docker images than anyone last couple months. I have setup build processes in most of the major CI services as has JC and at/for multiple types of organizations.

**Comment by adriancole** [comment]
here's first one that once polished can copy/pasta around https://github.com/hypertrace/pinot/pull/34

**Comment by JBAhire** [comment]
Docker CI checks have been added along with e2e tests in Hypertrace service and Ingester repo. So closing this now. Feel free to reopen if we are missing something.
